fms_extras/utils/generation.py

```diff
@@ -3,19 +3,564 @@
 from typing import Any, Callable, Dict, List, MutableMapping, Optional, Tuple, Union
 
 import torch
+import torch.nn as nn
 import torch.nn.functional as F
 
-from fms_extras.utils.cache.paged import PagedKVCacheManager
+from fms_extras.models.speculator import MLPSpeculator, apply_index_map, flatten_batch
+from fms_extras.utils.cache.paged import PagedAttentionCacheData, PagedKVCacheManager
 
 
-def __right_pad_zeros(input_tensor: torch.Tensor, max_length: int) -> torch.Tensor:
-    return torch.stack(
+def __execute_prefill(
+    model: Union[nn.Module, Callable],
+    input_ids: Union[torch.Tensor, List[torch.Tensor]],
+    cache_data: PagedAttentionCacheData,
+    max_seq_len: int,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Execute the prefill and return information needed for first prediction.
+
+    By prefill, we are referring to the initial forward pass on the prompt tokens prior
+    to any tokens being added to the cache
```

Comment:
is this the correct phrasing? it's equivalent to a kv-cache, so I don't think what goes in it is "tokens"? we're introducing terminology I don't think I've seen elsewhere, so we should make sure we're very clear what we mean.

Reply:
ill mention computed keys/values rather than tokens